File: app/src/main/java/com/tosh/poolandroid/View/RegisterActivity.java
package com.tosh.poolandroid.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.tosh.poolandroid.R;
import com.tosh.poolandroid.Retrofit.AuthRetrofitClient;
import com.tosh.poolandroid.Retrofit.NodeAuthService;
import com.tosh.poolandroid.ViewModel.RegistrationViewModel;

import io.reactivex.disposables.CompositeDisposable;
import retrofit2.Retrofit;

import static androidx.core.util.PatternsCompat.EMAIL_ADDRESS;

public class RegisterActivity extends AppCompatActivity {

    private MaterialButton buttonLogin;
    private MaterialButton buttonRegister;
    private TextInputEditText inputName, inputEmail, inputPassword, inputConfirmPassword;

    private NodeAuthService api;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    private SharedPreferences preferences;
    private SharedPreferences.Editor editor;
    private RegistrationViewModel registrationViewModel;
    private String email;
    private String name;
    private String password;
    private String confirmpassword;





    @Override
    protected void onStop() {
        compositeDisposable.clear();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        compositeDisposable.clear();
        super.onDestroy();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        //init api
        Retrofit retrofit = AuthRetrofitClient.getInstance();
        api = retrofit.create(NodeAuthService.class);

        // views
        buttonLogin = findViewById(R.id.register_login_btn);
        buttonRegister = findViewById(R.id.register_btn);
        inputName = findViewById(R.id.name_register);
        inputEmail = findViewById(R.id.email_register);
        inputPassword = findViewById(R.id.password_register);
        inputConfirmPassword = findViewById(R.id.confirm_password_register);

        buttonLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(RegisterActivity.this, LoginActivity.class);
                startActivity(intent);
                finish();
            }
        });

        buttonRegister.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                email = inputEmail.getText().toString();
                name = inputName.getText().toString();
                password = inputPassword.getText().toString();
                confirmpassword = inputConfirmPassword.getText().toString();

                registerUser(name, email, password, confirmpassword);
            }
        });

        instatiateRegisterViewModel();


    }

    private void registerUser(final String name, final String email, String password, String confirmPassword) {
        if(TextUtils.isEmpty(name)){
            inputName.setError("Enter name");
            return;
        }
        if(TextUtils.isEmpty(email)){
            inputEmail.setError("Enter email");
            return;
        }

        if(!EMAIL_ADDRESS.matcher(email).matches()){
            inputEmail.setError("Enter valid email");
            return;
        }
        if(TextUtils.isEmpty(password)){
            inputPassword.setError("Enter password");
            return;
        }
        if(TextUtils.isEmpty(confirmPassword)){
            inputConfirmPassword.setError("Confirm password");
            return;
        }
        if(!TextUtils.equals(password, confirmPassword)){
            inputConfirmPassword.setError("Passwords don't match");
            return;
        }

        if(password.length()< 6){
            inputPassword.setError("Password should be atleast 6 characters");
        }

        registrationViewModel.registerUser(name, email, password, confirmPassword);

    }

    private void instatiateRegisterViewModel(){
        registrationViewModel = ViewModelProviders.of(this).get(RegistrationViewModel.class);

        registrationViewModel.getRegistrationResult().observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                if (s.equals("success")){
                    addToSharedPreferences(email, name);
                    Intent intent = new Intent(RegisterActivity.this, PhoneActivity.class);
                    startActivity(intent);
                    finish();
                }else{
                    Toast.makeText(RegisterActivity.this, s, Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    public void addToSharedPreferences(String email, String name){
        preferences = PreferenceManager.getDefaultSharedPreferences(this);
        editor = preferences.edit();
        editor.putString("email", email);
        editor.putString("name", name);
        editor.apply();

    }
}

File: app/src/main/java/com/tosh/poolandroid/view/RegisterActivity.kt
package com.tosh.poolandroid.view

import android.content.Intent
import android.content.SharedPreferences
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.WindowManager
import android.widget.Toast
import androidx.core.util.PatternsCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.tosh.poolandroid.R
import com.tosh.poolandroid.util.show
import com.tosh.poolandroid.viewmodel.RegistrationViewModel
import kotlinx.android.synthetic.main.activity_register.*

class RegisterActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        this.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)

        register_login_btn.setOnClickListener {
            val intent = Intent(applicationContext, LoginActivity::class.java)

            startActivity(intent)
        }

        register_btn.setOnClickListener {
            val name = name_register.text.toString().trim()
            val email = email_register.text.toString().trim()
            val password = password_register.text.toString().trim()
            val confirmPassword = confirm_password_register.text.toString().trim()

            if (name.isEmpty()){
                name_register.error = "Enter name"
                return@setOnClickListener
            }
            if (email.isEmpty()){
                email_register.error = "Enter email"
                return@setOnClickListener
            }
            if (!PatternsCompat.EMAIL_ADDRESS.matcher(email).matches()) {
                email_register.error = "Enter a valid email"
                return@setOnClickListener
            }
            if (password.isEmpty()){
                password_register.error = "Enter password"
                return@setOnClickListener
            }

            if (password.length <6){
                password_register.error = "Password should be more than 6 characters long"
                return@setOnClickListener
            }

            if (!password.equals(confirmPassword)){
                confirm_password_register.error = "Passwords dont match"
                return@setOnClickListener
            }
            instantiateRegisterViewModel(name, email, password, confirmPassword)
        }
    }

    private fun instantiateRegisterViewModel(name:String, email:String,  password:String, confirmPassord:String) {
        val registrationViewModel: RegistrationViewModel = ViewModelProviders.of(this)[RegistrationViewModel::class.java]
        registrationViewModel.userRegister(name,email, password, confirmPassord).observe(this, Observer {

            if (it == "successful"){
                val intent = Intent(applicationContext, PhoneActivity::class.java)
                intent.putExtra("Name", name)
                intent.putExtra("Email", email)
                addToSharedPreferences(email)
                register_progress.show()
                startActivity(intent)
                finish()
            }else{
                Toast.makeText(applicationContext,it, Toast.LENGTH_LONG).show()
            }
        })
    }

    fun addToSharedPreferences(email: String) {
        var preferences: SharedPreferences?
        var editor: SharedPreferences.Editor?

        preferences = PreferenceManager.getDefaultSharedPreferences(this)
        editor = preferences.edit()
        editor.putString("email", email)
        editor.apply()

    }
}
